Close LGVISIUM-63: Extraction of the groundwater logo using computer vision

dcleres opened the pull request:
Addition of template matching to the code. This should make it possible to find groundwater information without the respective keywords. 

In this PR, I also tried to remove the keywords that were not used and to remove them if needed. Furthermore, I added a list of FP keywords. This list contains keywords that were leading to false positives in the detections.

stijnvermeeren-swisstopo reviewed (changes requested):
I would not enable this functionality by default at the moment, as it is just too slow and does not lead to an improvement that is significant enough to justify this slow-down. So maybe we should add a config parameter that can be used to decide whether to apply template matching or not. Then we can keep the code for now, and create a follow-up ticket to look into ways to optimize it.

One potential idea would be to look into using the OpenCV2 implementation of template matching instead of the scikit-learn one. Some people seem to claim that the former is more performant (e.g. https://www.reddit.com/r/opencv/comments/g8kdcs/question_the_speed_of_matchtemplate/).

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -137,9 +141,11 @@ class GroundwaterLevelExtractor(DataExtractor):
 
     # look for elevation values to the left, right and/or immediately below the key
-    search_left_factor: float = 2
+    search_left_factor: float = 3  # NOTE: check files 267125334-bp.pdf, 267125338-bp.pdf, and 267125339-bp.pdf if this
+    # value is too high, as it might lead to false positives
```
Yes, this, in combination with the new `search_above_factor`, indeed seems to lead to too many false positives. But maybe the ongoing work in https://jira.swisstopo.ch/browse/LGVISIUM-77 will already make this more robust again?

dcleres replied:
The issue I was facing with the 267125334-bp.pdf, 267125338-bp.pdf, and 267125339-bp.pdf bore profiles was that False Positives were generated if the left search factor was too large. In these profiles, the algorithm would find the depth column and extract data. 

I think the best option performance-wise would be to use the default values from the `main` branch.

dcleres commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -287,10 +476,18 @@ def extract_groundwater(self, terrain_elevation: Elevation | None) -> list[Groun
+                found_groundwater, confidence_list = self.get_groundwater_from_illustration(
+                    lines, page_number, terrain_elevation
+                )
+                logger.info("Confidence list: %s", confidence_list)
+                print("Confidence list: %s", confidence_list)
+                logger.info("Found groundwater from illustration on page %s: %s", page_number, found_groundwater)
+                print("Found groundwater from illustration on page %s: %s", page_number, found_groundwater)
```
Indeed. I adapted this.

dcleres commented on src/stratigraphy/groundwater/utility.py:
```diff
@@ -45,12 +45,15 @@ def extract_depth(text: str, max_depth: int) -> float | None:
+                    depth = None
+                else:
+                    break
+            except ValueError:
+                continue
```
The code was essentially testing whether the year in tha date has two digits or four digits and then converted it into a date based format. We would run into the scenario of raising an exeception if the date has two digits to indicate the year instead of four. Since this implementation was not clear I changed it to something that is easier to read.

dcleres replied:
@stijnvermeeren-swisstopo and @dcleres looked into it together, and we decided to roll back to the current version in the `main` branch and create a follow-up ticket (LGVISIUM-100) to address this issue.

dcleres commented:
I addressed the comments you raised @stijnvermeeren-swisstopo . Thank you very much for the review. I added the possibility of running the template matching on demand by editing the environment variable `IS_SEARCHING_GROUNDWATER_ILLUSTRATION`. 

Even when not running the template matching, the metrics were improved: 

![Screenshot 2024-10-14 at 14 28 38](https://github.com/user-attachments/assets/d16774c6-a349-4b5d-94af-d0bf1afea0cc)

main branch was run: `vaunted-mink-342`

stijnvermeeren-swisstopo commented on pyproject.toml:
```diff
@@ -35,6 +35,7 @@ dependencies = [
     "opencv-python-headless",
     "quads>=1.1.0",
     "numpy<2",
+    "scikit-image==0.24.0"
```
move to optional dependencies + add corresponding documentation to README

dcleres replied:
Done.

stijnvermeeren-swisstopo commented on src/stratigraphy/data_extractor/data_extractor.py:
```diff
@@ -63,6 +66,11 @@ def __init__(self, document: fitz.Document):
 
         self.doc = document
         self.feature_keys = read_params("matching_params.yml")[f"{self.feature_name}_keys"]
+        self.feature_fp_keys = (
+            read_params("matching_params.yml")[f"{self.feature_name}_fp_keys"]
+            if read_params("matching_params.yml")[f"{self.feature_name}_fp_keys"]
+            else []
+        )
```
is this the same as `read_params("matching_params.yml")[f"{self.feature_name}_fp_keys"] or []` ?

dcleres replied:
Yes, and shorter! Thanks!

stijnvermeeren-swisstopo commented on src/stratigraphy/data_extractor/data_extractor.py:
```diff
@@ -105,7 +113,15 @@ def find_feature_key(self, lines: list[TextLine], allowed_error_rate: float = 0.
                 if match:
-                    matches.add(line)
+                    # Make sure the key is not in the false positive list
+                    is_fp_key = False
+                    for fp_key in self.feature_fp_keys:
+                        if fp_key in line.text:
+                            is_fp_key = True
+                            break
+
+                    if not is_fp_key:
```
Is this the same as 
```
if all(not fp_key in line.text for fp_key in self.feature_fp_keys):
```

dcleres replied:
I used this experession as it seemed a bit easier to read for me: 

`if not any(fp_key in line.text for fp_key in self.feature_fp_keys):`

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -2,12 +2,16 @@
 
 import fitz
 import numpy as np
+import skimage as ski
```
Move all logic related to template matching (including this import) to a separate Python file, and only import this file when template matching is enabled.

dcleres replied:
Done.

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -219,6 +232,12 @@ def get_groundwater_info_from_lines(self, lines: list[TextLine], page: int) -> G
                         date = extracted_date
+                        matched_lines_rect.append(line.rect)
+                else:
+                    # in case several dates are present, we skip the other dates
```
Clearer description:
"If a second date is present in the lines around the groundwater key, then we skip this line, instead of potentially falsely extracting a depth value from the date."

dcleres replied:
Thank you!

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -269,6 +288,184 @@ def get_groundwater_info_from_lines(self, lines: list[TextLine], page: int) -> G
+                    top_left[0], top_left[1], top_left[0] + template.shape[1], top_left[1] + template.shape[0]
+                )
+
+                # remove the matched area from the result to avoid finding the same area again
```
```suggestion
                # remove the matched area from the template matching result to avoid finding the same area again for the same template
```

stijnvermeeren-swisstopo commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -285,12 +482,24 @@ def extract_groundwater(self, terrain_elevation: Elevation | None) -> list[Groun
+
+            found_groundwater = self.get_groundwater_near_key(lines, page_number)
+            if not found_groundwater:
+                logger.info("No groundwater found near the key on page %s.", page_number)
```
Not necessary to log this.

dcleres replied:
Removed the logging about this

dcleres commented on src/stratigraphy/groundwater/groundwater_extraction.py:
```diff
@@ -285,12 +482,24 @@ def extract_groundwater(self, terrain_elevation: Elevation | None) -> list[Groun
+                        lines, page_number, terrain_elevation
+                    )
+                    if found_groundwater:
+                        logger.info("Confidence list: %s", confidence_list)
```
Is this helpful logging according to you, @stijnvermeeren-swisstopo? In a previous iteration, we removed the logging in case no groundwater was found. We can also remove the logging when groundwater is found, as I mostly use it for debugging purposes.

dcleres commented:
@stijnvermeeren-swisstopo I do believe I implemented all the changes we discussed today. Main change, the template matching is now in an independent and separate file.

dcleres commented on pyproject.toml:
```diff
@@ -54,6 +54,9 @@ visualize = [
+    "scikit-image==0.24.0"
+]
 
 all = ["swissgeol-boreholes-dataextraction[test, lint, experiment-tracking, visualize, devtools]"]
```
Oh thanks! I was not aware of that!

dcleres commented:
@stijnvermeeren-swisstopo thank you for your review. I add the `groundwater_illustration_matching` to the toml installation script.

stijnvermeeren-swisstopo commented on pyproject.toml:
```diff
@@ -54,8 +54,11 @@ visualize = [
 devtools = [
     "tqdm"
 ]
+groundwater_illustration_matching = [
```
It still doesn't quite work for me to do install scikit-image with `pip install -e '.[all]'` .
Probably need to use hyphens here instead of underscores (i.e. `groundwater-illustration-matching`).
I think there's some normalisation happening in the background that interferes with underscores: see e.g. https://peps.python.org/pep-0685/
We had the same issue already once before with Renato: https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/54#discussion_r1625849873

dcleres replied:
Ah thank you for letting me know!

stijnvermeeren-swisstopo reviewed (approved):
LGTM!
